atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/sweep/CandidatePagingState.java

```diff
@@ -0,0 +1,180 @@
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.Preconditions;
+import com.palantir.atlasdb.encoding.PtBytes;
+import com.palantir.atlasdb.keyvalue.api.CandidateCellForSweeping;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.ImmutableCandidateCellForSweeping;
+import com.palantir.atlasdb.keyvalue.api.RangeRequests;
+
+import gnu.trove.list.TLongList;
+import gnu.trove.list.array.TLongArrayList;
+
+public class CandidatePagingState {
+    private byte[] currentRowName;
+    private byte[] currentColName = PtBytes.EMPTY_BYTE_ARRAY;
+    private Long maxSeenTimestampForCurrentCell = null;
+    private final TLongList currentCellTimestamps = new TLongArrayList();
+    private boolean currentIsLatestValueEmpty = false;
+    private long cellTsPairsExamined = 0;
+    private boolean reachedEnd = false;
+
+    private CandidatePagingState(byte[] currentRowName) {
+        this.currentRowName = currentRowName;
+    }
+
+    public static CandidatePagingState create(byte[] startRowInclusive) {
+        return new CandidatePagingState(startRowInclusive);
+    }
+
+    public static class StartingPosition {
+        public final byte[] rowName;
+        public final byte[] colName;
+        @Nullable public final Long timestamp;
+
+        public StartingPosition(byte[] rowName, byte[] colName, Long timestamp) {
+            this.rowName = rowName;
+            this.colName = colName;
+            this.timestamp = timestamp;
+        }
+    }
+
+    public static class BatchResult {
+        public final List<CandidateCellForSweeping> candidates;
+        public final boolean reachedEnd;
+
+        private BatchResult(List<CandidateCellForSweeping> candidates, boolean reachedEnd) {
+            this.candidates = candidates;
+            this.reachedEnd = reachedEnd;
+        }
+    }
+
+    public static class KvsEntryInfo {
+        public final byte[] rowName;
+        public final byte[] colName;
+        public final long ts;
+        public final boolean isEmptyValue;
+
+        public KvsEntryInfo(byte[] rowName, byte[] colName, long ts, boolean isEmptyValue) {
+            this.rowName = rowName;
+            this.colName = colName;
+            this.ts = ts;
+            this.isEmptyValue = isEmptyValue;
+        }
+    }
+
+    public BatchResult processBatch(List<KvsEntryInfo> cellTsPairs, int maxCellTsPairsExpected) {
+        Preconditions.checkArgument(maxCellTsPairsExpected > 0, "maxCellTsPairsExpected must be strictly positive");
+        List<CandidateCellForSweeping> candidates = new ArrayList<>();
+        for (KvsEntryInfo cellTs : cellTsPairs) {
+            checkCurrentCellAndUpdateIfNecessary(cellTs).ifPresent(candidates::add);
+            if (maxSeenTimestampForCurrentCell != null) {
+                // We expect the timestamps in ascending order. This check costs us a few CPU cycles
```

Comment:
Since this is a public method, it's probably worth explaining this in a javadoc for the method. I assume that in practice, we sort the timestamps before feeding them to `processBatch`.

Reply:
Public method in an internal implementation helper class. But sure, I'll add a comment